lib/thrift/generator/utils.ex

```diff
@@ -60,30 +60,47 @@ defmodule Thrift.Generator.Utils do
   end
 
   @doc """
-  Merge the binaries in an iolist.
+  Optimize an expression that returns an iolist.
 
-    ["a", "b", ["c", [var]]] => ["abc", var]
+    [<<1>>, <<2>>] => <<1, 2>>
   """
-  def merge_binaries([a | rest]) when is_list(a) do
-    merge_binaries(a ++ rest)
+  def optimize_iolist([a | rest]) when is_list(a) do
+    optimize_iolist(a ++ rest)
   end
-  def merge_binaries([a, b | rest]) when is_list(b) do
-    merge_binaries([a] ++ b ++ rest)
+  def optimize_iolist([a, b | rest]) when is_list(b) do
+    optimize_iolist([a] ++ b ++ rest)
   end
-  def merge_binaries([{:<<>>, [], a}, {:<<>>, [], b} | rest]) do
-    merge_binaries([{:<<>>, [], a ++ b} | rest])
+  def optimize_iolist([{:<<>>, opts, a}, {:<<>>, _, b} | rest]) do
+    optimize_iolist([{:<<>>, opts, a ++ b} | rest])
   end
-  def merge_binaries([a | rest]) do
-    [a] ++ merge_binaries(rest)
+  def optimize_iolist([a, {:<<>>, opts, b} | rest]) when is_binary(a) do
+    optimize_iolist([{:<<>>, opts, [a | b]} | rest])
   end
-  def merge_binaries(a) do
-    a
+  def optimize_iolist([{:<<>>, opts, a}, b | rest]) when is_binary(b) do
+    optimize_iolist([{:<<>>, opts, a ++ [b]} | rest])
   end
-
-  def simplify_iolist([{:<<>>, _, _} = binary]) do
+  def optimize_iolist([{:<<>>, opts, a}, <<0>> | rest]) do
+    optimize_iolist([{:<<>>, opts, a ++ [0]} | rest])
+  end
+  def optimize_iolist([{:|, _, [a, b]} | rest]) do
+    optimize_iolist([a, b | rest])
+  end
+  def optimize_iolist([a, b]) do
+    [{:|, [], [a, b]}]
```

Comment:
A list is represented in memory as a linked list of cons cells. A cons cell is a pair of adjacent values `[head | tail]` where `head` is the value at that point in the list and `tail` is either the next cons cell or the empty list. For example, a list with three elements looks like:

```["a" | ["b" | ["c" | []]]]```

An iolist is also composed out of cons cells but one less constraint: the tail can also be a binary. This means you can save one cons when constructing an iolist. The same list above could be converted to the iolist:

```["a" | ["b" | "c"]]```

It uses slightly less memory. It's also slightly faster to traverse, as there's one less pointer to follow. This optimization is performed by the above `optimize_iolist` clause.